prov/util/src/util_cq.c

```diff
@@ -570,66 +510,242 @@ void ofi_cq_progress(struct util_cq *cq)
+	.bind = fi_no_bind,
+	.control = fi_no_control,
+	.ops_open = fi_no_ops_open,
+};
+
+static int util_init_peer_cq(struct util_cq *cq, struct fi_cq_attr *attr)
+{
+	cq->peer_cq = calloc(1, sizeof(*cq->peer_cq));
+	if (!cq->peer_cq)
+		return -FI_ENOMEM;
+
+
+	slist_init(&cq->aux_queue);
 
 	switch (attr->format) {
 	case FI_CQ_FORMAT_UNSPEC:
 	case FI_CQ_FORMAT_CONTEXT:
-		read_func = util_cq_read_ctx;
+		cq->read_entry = util_cq_read_ctx;
 		break;
 	case FI_CQ_FORMAT_MSG:
-		read_func = util_cq_read_msg;
+		cq->read_entry = util_cq_read_msg;
 		break;
 	case FI_CQ_FORMAT_DATA:
-		read_func = util_cq_read_data;
+		cq->read_entry = util_cq_read_data;
 		break;
 	case FI_CQ_FORMAT_TAGGED:
-		read_func = util_cq_read_tagged;
+		cq->read_entry = util_cq_read_tagged;
 		break;
 	default:
 		assert(0);
 		return -FI_EINVAL;
 	}
 
-	ret = fi_cq_init(domain, attr, read_func, cq, context);
+	cq->cirq = util_comp_cirq_create(attr->size == 0 ? UTIL_DEF_CQ_SIZE : attr->size);
+	if (!cq->cirq)
+		return -FI_ENOMEM;
+
+	if (cq->domain->info_domain_caps & FI_SOURCE) {
+		cq->src = calloc(cq->cirq->size, sizeof(*cq->src));
+		if (!cq->src)
+			return -FI_ENOMEM;
+		cq->peer_cq->owner_ops = &util_peer_cq_src_owner_ops;
+	} else {
+		cq->peer_cq->owner_ops = &util_peer_cq_owner_ops;
+	}
+
+	cq->peer_cq->fid.fclass = FI_CLASS_PEER_CQ;
+	cq->peer_cq->fid.context = cq;
+	cq->peer_cq->fid.ops = &util_peer_cq_fi_ops;
+
+	return FI_SUCCESS;
+}
+
+int ofi_cq_init(const struct fi_provider *prov, struct fid_domain *domain,
+		 struct fi_cq_attr *attr, struct util_cq *cq,
+		 ofi_cq_progress_func progress, void *context)
+{
+	struct fi_wait_attr wait_attr;
+	struct fid_wait *wait;
+	enum ofi_lock_type lock_type;
+	int ret;
+
+	assert(progress);
+	ret = ofi_check_cq_attr(prov, attr);
 	if (ret)
 		return ret;
 
-	/* CQ must be fully operational before adding to wait set */
-	if (cq->wait) {
+	cq->cq_fid.fid.ops = &util_cq_fi_ops;
+	cq->cq_fid.ops = &util_cq_ops;
+	cq->progress = progress;
+
+	cq->domain = container_of(domain, struct util_domain, domain_fid);
+	ofi_atomic_initialize32(&cq->ref, 0);
+	ofi_atomic_initialize32(&cq->wakeup, 0);
+	dlist_init(&cq->ep_list);
+	ofi_mutex_init(&cq->ep_list_lock);
+
+	if (cq->domain->threading == FI_THREAD_COMPLETION ||
+	    cq->domain->threading == FI_THREAD_DOMAIN)
+		lock_type = OFI_LOCK_NOOP;
+	else
+		lock_type = cq->domain->lock.lock_type;
+
+	ret = ofi_genlock_init(&cq->cq_lock, lock_type);
+	if (ret)
+		return ret;
+
+	cq->flags = attr->flags;
+	cq->cq_fid.fid.fclass = FI_CLASS_CQ;
+	cq->cq_fid.fid.context = context;
+
+	switch (attr->wait_obj) {
+	case FI_WAIT_NONE:
+		wait = NULL;
+		break;
+	case FI_WAIT_UNSPEC:
+	case FI_WAIT_FD:
+	case FI_WAIT_POLLFD:
+	case FI_WAIT_MUTEX_COND:
+	case FI_WAIT_YIELD:
+		memset(&wait_attr, 0, sizeof wait_attr);
+		wait_attr.wait_obj = attr->wait_obj;
+		cq->internal_wait = 1;
+		ret = fi_wait_open(&cq->domain->fabric->fabric_fid,
+				   &wait_attr, &wait);
+		if (ret)
+			return ret;
```

Comment:
I'm not sure the cleanup in the error paths is correct.  We called ofi_genlock_init() above, but are exiting on error without calling ofi_genlock_destroy().  The prior code jumped to cleanup